### DISCUSSION TITLE
Exceeding 30 Second Lambda Limit / Async Dispatch

### DISCUSSION THREAD

**Discussion by LarsaSolidor** [stefanzweifel/sidecar-browsershot#86] [Q&A] [answered]
Hello, 

I'm trying to use Browsershot to render a few hundred pages of output, which takes longer than the 30 second lambda execution time available, as I understand it. 

Is there a way to set the Browsershot function off async, so it'll store the resulting PDF on S3, and it can have the full 15 minute lambda execution time in order to generate that PDF?

**Comment by stefanzweifel** [comment, marked as answer]
This is currently badly documented, but you can set the `SIDECAR_TIMEOUT`-env variable to change the timeout setting of the function. Don't forget to redeploy your function.
(This would be a global change that affects all Sidecar functions in your project; made a note to add a dedicated setting in v2 (coming later this year #85))

Or you could manually update the timeout setting in the AWS web UI. (Setting reset after each deployment though)

<img width="825" alt="Screenshot 2023-09-11 at 20 46 08" src="https://github.com/stefanzweifel/sidecar-browsershot/assets/1080923/6b8ff52b-87a7-4f79-880a-2e51d2799028">

---

There's currently no async-support in sidecar-browsershot. However, Sidecar supports async execution ([link](https://hammerstone.dev/sidecar/docs/main/functions/executing#sync-vs-async-vs-event)).

Not sure how that would work or how we could expose an API for this.
Currently sidecar-browsershot doesn't really expose any Sidecar _functions_ to you as a user. If we would start support async, we probably would have to expose the `PendingResult` to you, so you could interact or even store it somewhere.

Have to think about this more.


Currently you can store files directly on S3 ([docs](https://github.com/stefanzweifel/sidecar-browsershot#saving-directly-to-s3)), but that is a sync action and would take 15 minutes in your project.

**Reply by stefanzweifel** [reply]
This discussion and the request to support async Sidecar functions was still open in my todo app …

After re-reading the [Sidecar docs covering async functions](https://hammerstone.dev/sidecar/docs/main/functions/executing#sync-vs-async-vs-event) I've decided that this is currently out of scope for this package.

I can't see an easy and good way to support async functions, without rewriting huge parts of the package and how we utilize `spatie/browsershot` under the hood.
But nothing is stopping you or others to write their own `BrowsershotAsyncLambda` based on [`BrowsershotLambda`](https://github.com/stefanzweifel/sidecar-browsershot/blob/868cae707c018244c8149c7c133ffaf97dd87537/src/BrowsershotLambda.php).

Closing this discussion for now.

**Reply by stefanzweifel** [reply]
And regarding timeouts: A recent [release](https://github.com/stefanzweifel/sidecar-browsershot/releases/tag/v1.11.0) added a `SIDECAR_BROWSERSHOT_TIMEOUT` env variable, to define the timeout of the Browsershot Lambda function.
